sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp

```diff
@@ -0,0 +1,2630 @@
+        }
+        auto response_serverencrypted_iterator
+            = http_response.GetHeaders().find("x-ms-server-encrypted");
+        if (response_serverencrypted_iterator != http_response.GetHeaders().end())
+        {
+          response.ServerEncrypted = response_serverencrypted_iterator->second == "true";
+        }
+        auto response_encryptionkeysha256_iterator
+            = http_response.GetHeaders().find("x-ms-encryption-key-sha256");
+        if (response_encryptionkeysha256_iterator != http_response.GetHeaders().end())
+        {
+          response.EncryptionKeySHA256 = response_encryptionkeysha256_iterator->second;
+        }
+        return response;
+      }
+
+      static BlockInfo StageBlock(
+          Azure::Core::Context context,
+          Azure::Core::Http::HttpPipeline& pipeline,
+          const std::string& url,
+          const StageBlockOptions& options)
+      {
+        auto request = StageBlockConstructRequest(url, options);
+        auto response = pipeline.Send(context, request);
+        return StageBlockParseResponse(*response);
+      }
+
+      struct CommitBlockListOptions
+      {
+        std::string Version;
+        std::string Date;
+        std::vector<std::pair<BlockType, std::string>> BlockList;
+        BlobHTTPHeaders Properties;
+        std::map<std::string, std::string> Metadata;
+        std::string LeaseId;
+        std::string EncryptionKey;
+        std::string EncryptionKeySHA256;
+        std::string EncryptionAlgorithm;
+        AccessTier Tier = AccessTier::Unknown;
+      }; // struct CommitBlockListOptions
+
+      static Azure::Core::Http::Request CommitBlockListConstructRequest(
+          const std::string& url,
+          const CommitBlockListOptions& options)
+      {
+        // TODO: Think about how to initialize
+        // xmlInitParser();
+        // TODO: Think about how to free doc on exception
+        using namespace libXML2;
+        xmlDocPtr doc = xmlNewDoc(BAD_CAST("1.0"));
+        xmlNodePtr block_list_node = xmlNewNode(nullptr, BAD_CAST("BlockList"));
+        xmlDocSetRootElement(doc, block_list_node);
+
+        for (const auto& block : options.BlockList)
+        {
+          const char* tag_name = nullptr;
+          if (block.first == BlockType::Uncommitted)
+            tag_name = "Uncommitted";
+          else if (block.first == BlockType::Committed)
+            tag_name = "Committed";
+          else if (block.first == BlockType::Latest)
+            tag_name = "Latest";
+          else
+            throw std::runtime_error("unexpected block type");
+
+          xmlNewChild(block_list_node, nullptr, BAD_CAST(tag_name), BAD_CAST(block.second.data()));
+        }
+
+        xmlChar* xml_dump;
+        int xml_dump_size;
+        xmlDocDumpMemory(doc, &xml_dump, &xml_dump_size);
+        xmlFreeDoc(doc);
+        if (xml_dump == nullptr)
+        {
+          throw std::runtime_error("failed to allocate memory when building xml body");
```

Comment:
>runtime_error [](start = 21, length = 13)

Are we going to change this to StorageError/AzureError in the future?

Reply:
I think we should have a `StorageError` which inherits from `std::runtime_error`.

Reply:
But this is not priority.